include/drivers/i2c.h

```diff
@@ -212,6 +296,52 @@ static inline int z_impl_i2c_configure(struct device *dev, u32_t dev_config)
 	return api->configure(dev, dev_config);
 }
 
+/** @brief Perform asynchronous data transfer to another device.
+ *
+ * This routine provides a generic interface to perform asynchronous transfer of
+ * single message to another I2C device. Function can be called from interrupt
+ * context.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param msg Message to transfer.
+ * @param addr Address of the I2C target device.
+ * @param callback Callback called on transfer completion.
+ * @param user_data User data passed to the callback.
+ *
+ * @retval Non-negative on successful transfer scheduling.
+ * @retval -EBUSY if device is currently performing a transfer.
+ * @retval -EINVAL if any parameter is invalid.
+ */
+static inline int i2c_single_transfer(struct device *dev,
+					const struct i2c_msg *msg, u16_t addr,
+					i2c_transfer_callback_t callback,
+					void *user_data)
+{
+	const struct i2c_driver_api *api =
+		(const struct i2c_driver_api *)dev->driver_api;
+
+	return api->single_transfer(dev, msg, addr, callback, user_data);
+}
+
+/** @brief Synchronous transfer using i2c_async API.
+ */
+int z_i2c_async_sync_transfer(struct device *dev, struct i2c_msg *msgs,
+			    u8_t num_msgs, u16_t addr);
+
+/** @brief Get handle to the object for asynchronous transfers.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @return Asynchronous object or NULL if not supported.
+ */
+static inline struct i2c_async *i2c_get_async(struct device *dev)
```

Comment:
I see why you are doing this: it's easier then to manage async object memory space, as each and every driver would own it directly. But on the other hand, it is an object the driver will _never_ use by itself.
I think this needs to be reconsidered, finding a way to manager async objects directly into the async manager.

Reply:
i agree. currently i had to do something to get this object. I already made a proposal of different approach: #22409